File: app/src/main/java/soup/movie/ui/BasePresenter.java
package soup.movie.ui;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public abstract class BasePresenter<View extends BaseContract.View>
        implements BaseContract.Presenter<View> {

    protected View view;

    private CompositeDisposable subscriptions;

    @Override
    public void attach(View view) {
        this.view = view;
        subscriptions = new CompositeDisposable();
        initObservable(subscriptions);
    }

    @Override
    public void detach() {
        subscriptions.dispose();
        subscriptions = null;
        this.view = null;
    }

    protected void initObservable(CompositeDisposable subscriptions) {
    }

    protected final void register(Disposable disposable) {
        subscriptions.add(disposable);
    }
}

File: app/src/main/java/soup/movie/ui/BasePresenter.kt
package soup.movie.ui

import androidx.annotation.CallSuper
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.internal.disposables.DisposableContainer
import soup.movie.ui.BaseContract.Presenter
import soup.movie.ui.BaseContract.View

abstract class BasePresenter<V : View> : Presenter<V> {

    protected var view: V? = null

    private var subscriptions: CompositeDisposable? = null

    override fun onAttach(view: V) {
        this.view = view
        subscriptions = CompositeDisposable()
                .also { initObservable(it) }
    }

    override fun onDetach() {
        subscriptions?.dispose()
        subscriptions = null
        this.view = null
    }

    @CallSuper
    protected open fun initObservable(disposable: DisposableContainer) {
        //stub implementation
    }
}
